File: addons/terrabrush/Scripts/Tools/FoliageTool.cs
#if TOOLS
using Godot;

namespace TerraBrush;

public class FoliageTool : ToolBase {
    public FoliageTool(TerraBrush terraBrush) : base(terraBrush) {}

    protected override ImageTexture GetToolCurrentImageTexture(ZoneResource zone) {
        return zone.FoliagesTexture[_terraBrush.FoliageIndex.Value];
    }

    public override void Paint(TerrainToolType toolType, Image brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
        if (_terraBrush.FoliageIndex == null) {
            return;
        }

        ForEachBrushPixel(brushImage, brushSize, imagePosition, (imageZoneInfo, pixelBrushStrength) => {
            var currentPixel = imageZoneInfo.Image.GetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y);
            var newColor = toolType == TerrainToolType.FoliageAdd ? Colors.Red : new Color(0, 0, 0, 0);

            var newValue = new Color(
                Mathf.Lerp(currentPixel.R, newColor.R, pixelBrushStrength * brushStrength),
                Mathf.Lerp(currentPixel.G, newColor.G, pixelBrushStrength * brushStrength),
                Mathf.Lerp(currentPixel.B, newColor.B, pixelBrushStrength * brushStrength),
                Mathf.Lerp(currentPixel.A, newColor.A, pixelBrushStrength * brushStrength)
            );
            imageZoneInfo.Image.SetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y, newValue);
        });

        _terraBrush.TerrainZones.UpdateFoliagesTextures(_terraBrush.FoliageIndex.Value);
    }
}
#endif

File: src/editor_tools/foliage_tool.cpp
#include "foliage_tool.h"

using namespace godot;

void FoliageTool::_bind_methods() {}

FoliageTool::FoliageTool() {}

FoliageTool::~FoliageTool() {}

Ref<ImageTexture> FoliageTool::getToolCurrentImageTexture(Ref<ZoneResource> zone) {
    return zone->get_foliagesTexture()[_selectedFoliageIndex];
}

void FoliageTool::paint(TerrainToolType toolType, Ref<Image> brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
    if (_selectedFoliageIndex < 0) {
        return;
    }

    forEachBrushPixel(brushImage, brushSize, imagePosition, ([&](ImageZoneInfo imageZoneInfo, float pixelBrushStrength) {
        Color currentPixel = imageZoneInfo.image->get_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y);
        Color newColor = toolType == TerrainToolType::TERRAINTOOLTYPE_FOLIAGEADD ? Color(1, 0, 0, 1) : Color(0, 0, 0, 0);

        Color newValue = Color(
            Math::lerp(currentPixel.r, newColor.r, pixelBrushStrength * brushStrength),
            Math::lerp(currentPixel.g, newColor.g, pixelBrushStrength * brushStrength),
            Math::lerp(currentPixel.b, newColor.b, pixelBrushStrength * brushStrength),
            Math::lerp(currentPixel.a, newColor.a, pixelBrushStrength * brushStrength)
        );
        imageZoneInfo.image->set_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y, newValue);
    }));

    _terraBrush->get_terrainZones()->updateFoliagesTextures(_selectedFoliageIndex);
}

void FoliageTool::updateSelectedFoliageIndex(int value) {
    _selectedFoliageIndex = value;
}